mkosi/__init__.py

```diff
@@ -543,6 +549,157 @@ def install_systemd_boot(state: MkosiState) -> None:
+    for d in ("usr/lib/grub/i386-pc", "usr/share/grub2/i386-pc"):
+        if (p := state.root / d).exists() and any(p.iterdir()):
+            return p
+
+    return None
+
+
+def find_grub_install(state: MkosiState) -> Optional[Path]:
+    path = ":".join(os.fspath(p) for p in [state.root / "usr/bin", state.root / "usr/sbin"])
+
+    install = shutil.which("grub2-install", path=path) or shutil.which("grub-install", path=path)
+    if not install:
+        return None
+
+    return Path(install).relative_to(state.root)
+
+
+def want_grub_bios(state: MkosiState, partitions: Sequence[Partition] = ()) -> bool:
+    if state.config.bootable == ConfigFeature.disabled:
+        return False
+
+    if state.config.output_format != OutputFormat.disk:
+        return False
+
+    if state.config.bios_bootloader == BiosBootloader.none:
+        return False
+
+    have = find_grub_bios_directory(state) is not None
+    if not have and state.config.bios_bootloader == BiosBootloader.grub:
+        die("A BIOS bootable image with grub was requested but grub for BIOS is not installed")
+
+    bios = any(p.type == "21686148-6449-6e6f-744e-656564454649" for p in partitions)
+    if partitions and not bios and state.config.bios_bootloader == BiosBootloader.grub:
+        die("A BIOS bootable image with grub was requested but no BIOS Boot Partition was configured")
+
+    root = any(p.type.startswith("root") or p.type.startswith("usr") for p in partitions)
+    if partitions and not root and state.config.bios_bootloader == BiosBootloader.grub:
+        die("A BIOS bootable image with grub was requested but no root or usr partition was configured")
+
+    install = find_grub_install(state) is not None
+    if not install and state.config.bios_bootloader == BiosBootloader.grub:
+        die("A BIOS bootable image with grub was requested but grub-install was not found")
+
+    want = (have and bios and root and install) if partitions else have
+
+    control = Path("/dev/loop-control")
+    if not control.exists() or control.stat().st_uid != os.getuid():
+        if state.config.bios_bootloader == BiosBootloader.grub:
+            die("A BIOS bootable image with grub was requested but grub cannot be installed because mkosi is running in a user namespace")
+        if want:
+            logging.warning("Cannot install grub because mkosi is running in a user namespace, skipping")
+        return False
+
+    return want
+
+
+def prepare_grub_bios(state: MkosiState, partitions: Sequence[Partition]) -> None:
+    if not want_grub_bios(state, partitions):
+        return
+
+    root = finalize_roothash(partitions)
+    if not root:
+        root = next((f"root=PARTUUID={p.uuid}" for p in partitions if p.type.startswith("root")), None)
+    if not root:
+        root = next((f"mount.usr=PARTUUID={p.uuid}" for p in partitions if p.type.startswith("usr")), None)
+
+    assert root
+    assert (install := find_grub_install(state))
```

Comment:
Yes but we're not relying on the runtime behavior of assert here, I'm solely using it to refine the type because we know find_grub_install() will not return `None` here. So for that use it should be fine?

Reply:
I see. It definitely is more succinct, but it is a bit of a foot gun, because a stray `PYTHONOPTIMIZE=1` in the environment could remove the assert and then `install` would be unbound. It's more verbose, but
```python
install = find_grub_install(state)
if install is not None:
    die("Oh noes!")
```
will do the same type narrowing.

Unexpectedly, at least for me, since I thought the scoping rules would be different, even this seems to work:
```python
if (install := find_grub_install(state)) is not None:
    die("Oh noes!")
```

I just checked this for a smaller example, though:
```python
#!/usr/bin/python3
from typing import Optional

def foo() -> Optional[int]:
    return 5


def main():
    if (bar := foo()) is None:
        print("eh")
    print(bar)


if __name__ == "__main__":
    main()
```
which prints 5.

Anyway, in this case avoiding the foot gun is only one more line.

Reply:
So I just moved the assert to the next line, that way the variable will always be bound. the assert is just there to tell the type checker that the type is not None. If it is, we'll just get an exception later and it'll be a bug for us to fix

Reply:
Yeah, that's fine.